dev-docs/yaml-design-guidance.md

```diff
@@ -0,0 +1,293 @@
+# Yaml config design
+
+The router uses yaml configuration, and when creating new features or extending existing features you'll likely need to think about how configuration is exposed.
+
+In general users should have a pretty good idea of what a configuration option does without referring to the documentation.
+
+## Migrations
+
+We won't always get things right, and sometimes we'll need to provide [migrations](apollo-router/src/configuration/migrations/README.md) from old config to new config.
+
+Make sure you:
+1. Mention the change in the changelog
+2. Update docs
+3. Update any test configuration
+4. Create a migration test as detailed in [migrations](apollo-router/src/configuration/migrations/README.md)
+5. In your migration description tell the users what they have to update.
+
+## Process
+It should be obvious to the user what they are configuring and how it will affect Router behaviour. It's tricky for us as developers to know when something isn't obvious to users as often we are too close to the domain.
+
+Complex config should be run by the rest of the team. Ideally before writing any code, as coming to the team late can cause code churn and frustration. The process is as follows:
```

Comment:
Maybe let's have a less negative outlook here

```suggestion
Complex configuration changes should be discussed with the team before starting the implementation, since they will drive the code's design. The process is as follows:
```